rskj-core/src/main/java/org/ethereum/db/IndexedBlockStore.java

```diff
@@ -276,37 +304,65 @@ public synchronized Block getChainBlockByNumber(long number){
         return null;
     }
 
+    /**
+     * Retrieves a block header by its hash value.
+     *
+     * @param hash The hash of the block to look up, must be a valid Keccak256 byte array key.
+     *
+     * @return The block header if found, null if not.
+     */
     @Override
-    public synchronized Block getBlockByHash(byte[] hash) {
+    public synchronized BlockHeader getBlockHeaderByHash(byte[] hash) {
+        Keccak256 blockHeaderHash = new Keccak256(hash);
+        Optional<BlockHeader> cachedHeader = blockStoreCache.getBlockHeaderByHash(blockHeaderHash);
+        if (cachedHeader.isPresent()) {
+            return cachedHeader.get();
+        }
 
-        Block block = getBlock(hash);
-        if (block == null) {
+        Optional<BlockHeader> blockHeader = getDSValue(hash).flatMap(blockStoreEncoder::decodeBlockHeader);
+        blockHeader.ifPresent(blockStoreCache::addBlockHeader);
+        return blockHeader.orElse(null);
+    }
+
+
+    @Override
+    public synchronized Block getBlockByHash(byte[] hash) {
+        Optional<Block> optionalBlock = getBlock(hash);
+        if (!optionalBlock.isPresent()) {
             return null;
         }
 
-        blockCache.addBlock(block);
+        Block block = optionalBlock.get();
+        blockStoreCache.addBlock(block);
         remascCache.put(block.getHash(), getSiblingsFromBlock(block));
         return block;
     }
 
-    private synchronized Block getBlock(byte[] hash) {
-        Block block = this.blockCache.getBlockByHash(hash);
-
-        if (block != null) {
-            return block;
-        }
-
-        byte[] blockRlp = blocks.get(hash);
-        if (blockRlp == null) {
-            return null;
-        }
+    /**
+     * Retrieves a block from the cache or the data source.
+     *
+     * @param hash The hash too look up the block with, cannot be null.
+     * @return A block optional, empty if not found.
+     */
+    private synchronized Optional<Block> getBlock(byte[] hash) {
+        Optional<Block> cacheValue = this.blockStoreCache.getBlockByHash(new Keccak256(hash));
+        return cacheValue.isPresent() ? cacheValue : getDSValue(hash).flatMap(blockStoreEncoder::decodeBlock);
+    }
 
-        return blockFactory.decodeBlock(blockRlp);
+    /**
+     * Wraps in an optional the retrieved bytes value from the key value data source.
+     * This method should be used instead of accessing the data source directly.
+     *
+     * @param hash The key to look up the stored values, cannot be null.
+     * @return The wrapped key value response.
+     */
+    private Optional<byte[]> getDSValue(byte[] hash) {
```

Comment:
Receive a `Keccak256` instance instead of a `byte[]`. Always work with high-level concepts when possible.